### PR TITLE
Add FPC Load Balancer

### PR DESCRIPTION
**Pull request by style95**
## Description
This is to add FPCPoolBalancer
I opened a new PR to see if there is any difference in the Travis CI environment.
Old one: https://github.com/apache/openwhisk/pull/5158

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [x] Loadbalancer
- [ ] Scheduler
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by style95** [comment]
It's ready to review.

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+  override def activeActivationsByInvoker(invoker: String): Future[Int] = Future.successful(0)
+
+  /** Gets the number of in-flight activations in the system. */
+  override def totalActiveActivations: Future[Int] = Future.successful(totalActivations.intValue())
+
+  /** Gets the throttling for given action. */
+  override def checkThrottle(namespace: EntityPath, fqn: String): Boolean = {
+
+    /**
+     * Note! The throttle key is assumed to exist unconditionally and is not limited to throttle if not present.
```
Comment (style95) [inline core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:624]: Throttling works as follows.
Now the unit of throttling is the number of containers.
And there are two kinds of throttling, 1. `namespace throttling` and 2. `action throttling` with one configuration, `concurrent limit`.
Both throttlings are enabled under the following scenario.

Let's say the concurrent limit of a namespace is 30.
1. When there are 30 containers running in the namespace, the `namespace throttling` is enabled.
2. When the `namespace throttling` is enabled, no more new type of action is accepted.
   1. For example, if `actionA` is invoked and there are 30 containers for `actionA` only activations for `actionA` are allowed.
   2. When another `actionB` is invoked, it is rejected with 429 too many requests.
   3. This is because it's not possible to create more containers in the namespace
   4. When there are two actions `actionA` and `actionB` are running, activations for only those two actions are accepted in the same way.
3. If the workload keeps increasing, activations for `actionA` will be piled up in a queue because there are already 30 containers running and no more containers can be created.
4. If too many activations are buffered in a queue, `action throttling` is enabled. This means that the current containers cannot handle all activations in time. This threshold is configurable.
   1. There are two action throttling configurations, [max retention size](https://github.com/apache/openwhisk/blob/master/core/scheduler/src/main/resources/application.conf#L70), and [throttling fraction](https://github.com/apache/openwhisk/blob/master/core/scheduler/src/main/resources/application.conf#L72).
   2. If there are more activations in a queue than max retention size, action throttling is enabled.
   3. If there are fewer activations than `max-retention-size * throttling fraction`), action throttling is disabled.
   (This is to prevent too frequent throttling changes.)
5. So when a new activation comes, the load balancer checks whether the activation is throttled or not in this way.
   1. Check if `action throttling` is enabled, if yes reject the requests.
   2. If `action throttling` is not enabled, accept the requests.
   3. If `action throttling information` does not exist, checks `namespace throttling`. It means no queue is created for the action.
   4. If `namespace throttling` is enabled, reject the activation. This is because there is no container running and the namespace is already throttled.
   5. If `namespace throttling` is disabled or no information exists, it accepts the activation.

Comment (bdoyle0182) [reply core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:624]: For action throttling, is there no longer any concept of invocations per minute throttling? per minute throttling was useful for user applications to throttle to protect their downstream services that their action calls. Now it seems like there's no way to throttle based on what the action application can handle, but is strictly only for system throttling to protect openwhisk

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+  private val throttlingKey = ThrottlingKeys.prefix
+  private val watchedKeys = Set(queueKey, schedulerKey, throttlingKey)
+
+  private val watcher = actorSystem.actorOf(Props(new Actor {
+    watchedKeys.foreach { key =>
+      watcherService ! WatchEndpoint(key, "", true, watcherName, Set(PutEvent, DeleteEvent))
+    }
+
+    override def receive: Receive = {
+      case WatchEndpointRemoved(watchKey, key, value, true) =>
```
[the remark this answers is not in the document]
Comment (bdoyle0182) [reply core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:399]: Is it possible to get out of sync here? What if a queue is removed by the scheduler or the scheduler is brought down and this message isn't delivered to the load balancer? Can that happen? And if it can would that cause problems?

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+                .find(entry =>
+                  //the leader key's value is queue/invocationNamespace/ns/pkg/act/leader
+                  QueueKeys
+                    .queue(entry.invocationNamespace, entry.fullyQualifiedEntityName.copy(version = None), true) == key)
+                .foreach {
+                  entry =>
+                    implicit val transid = entry.transactionId
+                    logging.warn(
+                      this,
+                      s"The $key is deleted from ETCD, but there are still unhandled activations for this action, try to create a new queue")
```
Comment (bdoyle0182) [inline core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:414]: Trying to remember the architecture here, but how do unhandled activations still get retried after creating the new queue. Is it because this load balancer puts them on kafka and they'd just be sitting there waiting to be consumed by the scheduler with the recreated queue still? If the etcd queue gets deleted with an activation in it, then that's just lost at that point right since it's already been read off kafka?

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (style95) [reply core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:399]: These endpoints are kept alive by the lease.
When any components go wrong and cannot send a keepalive request for a given time(10s by default), data are automatically gone and controllers receive this event.

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala (hunk 1 above)
Comment (style95) [reply core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:624]: Yes. there is no such thing.
If that is necessary, we can add a new feature to the scheduler part. I believe we can control the rate activations are pulled out of a queue.
And if it is critical and urgent, we can indirectly limit invocations with the number of containers with the assumption that one action's execution time does not drastically vary.

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala (hunk 3 above)
Comment (style95) [reply core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:414]: When there is an activation in a queue, the queue is not deleted.
Unhandled activations here come from the possibility that while the queue is deleted because of timeout(no more activation), still there can be new activations incoming on the controller side.

Regarding the role of a controller, yes, now it is rather passive.
It just selects the right scheduler and send a queue creation request and put activations to one of the scheduler's Kafka topic.
Still, activations can be delivered to a scheduler without a queue for the given activations, then the scheduler looks for(with retries) the target scheduler that has a queue and forwards activations to the scheduler.
(I hope we can remove Kafka from the critical path entirely in the future. Then we are probably able to consolidate these two calls into one.)

So basically the situation here is supposed to rarely happen.
Once a queue is created, the activation will be forwarded to the proper scheduler.

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+
+  /**
+   * Assign an invoker to a message
+   *
+   * @param invokers Invoker pool
+   * @param minMemory Minimum memory for all invokers
+   * @return Assigned an invoker
+   */
+  // TODO add input/output example
+  def schedule(invokers: IndexedSeq[InvokerHealth], minMemory: ByteSize): Option[InvokerInstanceId] = {
```
[the remark this answers is not in the document]
Comment (style95) [reply core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala]: Oh.. this should be deleted.
This is only required for our downstream to send a message to debugging invokers.

**Review by bdoyle0182 — APPROVED** [review]

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala (hunk 1 above)
Comment (bdoyle0182) [reply core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:624]: I'm really doing a deep dive into the code and I'm a little confused on namespace throttling and how you prevent starvation of other actions within a namespace. As you explain, if `actionA` has all of the containers for a namespace and is at the limit but the namespace also wants to run some of `actionB` they just get throttled? If `actionA` sends consistent load to keep up with the container limit for the namespace but not backlog the queue to turn on action throttling, how would `actionB` ever get to run any instances? On the old scheduling algorithm it's just random based on how many activations in flight so even if you're hitting the concurrent throttle, you don't have to worry about one action completely starving another action in the namespace. Am I understanding this correctly?

Comment (style95) [reply core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala:624]: I think in such a case, it is natural to increase the namespace limit.
If someone regularly invokes an action utilizing all the resources and tries to run another one, and if both of the actions are supposed to run well he needs a higher resource limit. In the above case, there are not enough resources and I feel it's not that weird to get a `429 Too many requests` response.

When at least one container is created for `actionB` before the namespace limit is reached, `actionB` can still run as expected even after the namespace is throttled. However, there are not enough containers and activations will get piled up in queues, and eventually, the system will reject the requests with 429 responses if requests keep incoming at the same rate.

To be able to run both of them well, it is necessary to increase the limit to accommodate all of them.
Even if we increase the limit, it would be fine because the system doesn't create more containers than required.
